### ISSUE TITLE
Enable offline cache warming 

### ISSUE THREAD

**Issue by ramevads** [sothawo/mapjfx#64]
At the moment we are trying to download a specific region so that we can use the mapjfx module Offline within that region.

For that purpose, we would need a method for downloading tile URLs.  Such method would download in the cache the png and 'datainfo' file corresponding to the tile provided by the URL.

It would be nice to have that method working in the mapjfx module.
Many thanks!

**Comment by sothawo** [comment]
**Info**: I have it implemented on my Mac for Java 8 and 12, need to test it tomorrow on Windows before releasing it

**Comment by sothawo** [comment]
https://www.sothawo.com/2019/10/mapjfx-2-9-0-and-1-28-0-with-offline-cache-improvements/

**Comment by ramevads** [comment]
Hi Peter,
We are using the function addUrls over an specific area (aprox. 10x10 km) and we discovered that the first 800 patches are downloaded very quickly but after a few seconds the download speed drops dramatically for the remaining patches. Is the function addUrls  limiting the download speed after a specific threshold? Or is the OpenStreetMaps server limiting the speed?
Many thanks

**Comment by sothawo** [comment]
Hello Ramon,

haven’t thought about that. This is not a limit in MapView, but rather coming from OpenStreetmap. If you search for „openstreetmap throttling“ you will find that there are some limitations for downloading tiles. One is that a USER_AGENT must be sent - which I implemented - the other is throttling. I found some docs that limits to two threads in parallel. Currently I use a custom threadpool with the parallelism set to the number of processors. So we fire too many parallel requests and thus get throttled.

I think I should add an additional argument to the preload function that specifies the number of concurrent downloads, 0 meaning the current behaviour, then you could limit this to 1 or 2, and other users might use different values for different tile providers.

But I won’t find the time to do this before friday or saturday

**Comment by sothawo** [comment]
versions 2.10.1 and 1.29.1 are uploaded to Sonatype Nexus, should be in maven central soon.
There is a new method for preloading that takes an additional _parallelism_ argument, you should set that to at most 2 for OSM. Setting it below or equal to 0 will make the loading use the number of processors as before.
